### PR TITLE
Hessian refactor

### PR DESCRIPTION
**Pull request by reuvenperetz**
HessianServiceInfo is added and initiated during the core run.
It allows fetching hessian-based approximations based on a TraceHessianRequest that is sent. The TraceHessianRequest contains three parameters: The node to compute its approximation, the mode (weights/activations), which indicates whether to compute the Hessian info w.r.t weights/activations, and HessianInfoGranularity which indicates the number of scores to compute for this node (per-tensor, per-channel, per-element).
The service initiates the framework calculator (using the framework implementation it holds) and computes the hessian info based on the representative dataset it holds and the number of iterations to use for the approximation.
After computing the approximation, it caches the approximation in case this request will be requested again to avoid re-computation.

Additional changes:
The hessian service does not support graphs with non-compatible outputs like argmax, etc.
The score for a layer with multiple outputs is the average of the scores (one per output).

### CODE REVIEW

File: model_compression_toolkit/gptq/common/gptq_training.py [hunk 1]
```diff
@@ -132,25 +144,36 @@ def compute_hessian_based_weights(self,
...
-                                                              self.graph_float.get_inputs()},
-                                                             self.compare_points,
-                                                             output_list=model_output_replacement,
-                                                             all_outputs_indices=[],
-                                                             alpha=0,
-                                                             norm_weights=self.gptq_config.hessian_weights_config.norm_weights,
-                                                             n_iter=self.gptq_config.hessian_weights_config.hessians_n_iter)
+
+            # Use existing computations for the config to save time
+            existing_computations = hessian_service.fetch_hessian(self.gptq_hessian_config)
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/common/gptq_training.py:154]: Seems duplicated with the code in sensitivity evaluation, maybe put in an external utility method under hessian package

File: model_compression_toolkit/gptq/common/gptq_training.py [hunk 2]
```diff
@@ -132,25 +144,36 @@ def compute_hessian_based_weights(self,
...
+            for image_id, node_to_score in existing_computations.items():
+                image_ip_gradients = []
+                for ip in self.compare_points:
+                    image_ip_gradients.append(node_to_score[ip])
                 points_apprx_jacobians_weights.append(image_ip_gradients)
+
+            if images_to_compute_count > 0:
+                newly_computed = 0
+                for i in range(1, images.shape[0] + 1):
+                    Logger.info(f"Computing Jacobian-based weights approximation for image sample {i} out of {images.shape[0]}...")
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/common/gptq_training.py:164]: I know that this is copied, but while we are at it, replace "Jacobians" with "Hessian"

File: model_compression_toolkit/gptq/common/gptq_training.py [hunk 3]
```diff
@@ -132,25 +144,36 @@ def compute_hessian_based_weights(self,
...
+                    images = [images[i - 1:i]] * len(self.graph_float.get_inputs())  # TODO: change to image per input
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/common/gptq_training.py:165]: Same regarding assertion and exception

File: model_compression_toolkit/gptq/common/gptq_training.py [hunk 4]
```diff
@@ -132,25 +144,36 @@ def compute_hessian_based_weights(self,
...
+                    hessian_data = hessian_service.fetch_hessian(self.gptq_hessian_config,
+                                                                 images[i - 1:i])
+                    image_ip_gradients = []
+                    for ip in self.compare_points:
+                        image_ip_gradients.append(hessian_data[ip])
+                    points_apprx_jacobians_weights.append(image_ip_gradients)
```
Comment (ofirgo) [inline model_compression_toolkit/gptq/common/gptq_training.py:171]: Same regarding "jacobians" (change to "hessian").
Also, regarding the term "weights", let's decide on a different term and change it here as well.

File: model_compression_toolkit/core/keras/hessian/activation_hessian_calculator_keras.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+        the calculator.
+
+        Returns: Dictionary from interest point to hessian score.
+
+        """
+        if self.config.granularity == HessianGranularity.PER_LAYER:
+            output_list = self._get_model_output_replacement()
+            all_outputs_indices = []
+            if self.config.search_output_replacement:
+                all_outputs_indices = self._update_ips_with_outputs_replacements(output_list,
```
Comment (ofirgo) [inline model_compression_toolkit/core/keras/hessian/activation_hessian_calculator_keras.py:55]: GPTQ doesn't do this output_replacement hack, I think this change would make it run this too, and I don't know what would it do to the GPTQ execution.
Because we are planning on removing it, maybe for now just add a flag to skip this part when calling from GPTQ.

File: model_compression_toolkit/core/runner.py [hunk 6]
```diff
@@ -45,7 +46,7 @@
     WeightsFinalBitwidthConfigVisualizer, \
     ActivationFinalBitwidthConfigVisualizer
 from model_compression_toolkit.core.common.visualization.tensorboard_writer import TensorboardWriter
-
+from model_compression_toolkit.core.common.hessian import hessian_service
```
Comment (ofirgo) [inline model_compression_toolkit/core/runner.py:49]: Following the comment on the hessian_service variable, I think it would be better to initiate it here inside the runner instead of globally outside the service class.

File: model_compression_toolkit/core/runner.py [hunk 7]
```diff
@@ -54,7 +55,8 @@ def core_runner(in_model: Any,
                 fw_impl: FrameworkImplementation,
                 tpc: TargetPlatformCapabilities,
                 target_kpi: KPI = None,
-                tb_w: TensorboardWriter = None):
+                tb_w: TensorboardWriter = None,
+                additional_hessian_configs: List[HessianConfig]=None):
```
Comment (ofirgo) [inline model_compression_toolkit/core/runner.py:59]: Missing in function description (also, why "additional"?)

File: tests/keras_tests/function_tests/test_tensorboard_writer.py [hunk 8]
```diff
@@ -102,14 +104,19 @@ def plot_tensor_sizes(self):
                                          (4, 8), (4, 4), (4, 2),
                                          (2, 8), (2, 4), (2, 2)])
         tpc = generate_keras_tpc(name='mp_keras_tpc', tp_model=tpc_model)
+
+        # Hessian manager assumes core should be initialized. This test does not do it, so we disable the use of hessians in MP
```
Comment (ofirgo) [inline tests/keras_tests/function_tests/test_tensorboard_writer.py:108]: Hessian service

**Review by ofirgo — CHANGES_REQUESTED** [review]
General comment: let's discuss offline about whether or not we are planning on creating a "hessian stage" in the runner that would enable all the hessian computation (I thought that this is part of this PR, if not that's ok, just want to make sure what the plan is)

File: model_compression_toolkit/core/common/hessian/hessian_service.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+                                                       input_images=input_images,
+                                                       fw_impl=self.fw_impl)
+            hessian = hessian_calculator.compute()
+
+        if hessian_cfg in self.hessian_cfg_to_hessian_data:
+            self.hessian_cfg_to_hessian_data[hessian_cfg][id(input_images)] = hessian
+        else:
+            self.hessian_cfg_to_hessian_data[hessian_cfg] = {id(input_images): hessian}
+
+    def fetch_hessian(self,
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/hessian/hessian_service.py:89]: split to two functions - all and for single input image

Comment (reuvenperetz) [reply model_compression_toolkit/core/common/hessian/hessian_service.py]: It's been replaced with a method to fetch a number of approximations.

File: model_compression_toolkit/core/keras/hessian/hessian_calculator_keras.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
+from typing import Dict, List
+
+from model_compression_toolkit.core.common import Graph
+from model_compression_toolkit.core.common.hessian.hessian_calculator import HessianCalculator
+import tensorflow as tf
+
+from model_compression_toolkit.core.common.hessian.hessian_config import HessianConfig
+
+class HessianCalculatorKeras(HessianCalculator):
```
[the remark this answers is not in the document]
Comment (reuvenperetz) [reply model_compression_toolkit/core/keras/hessian/hessian_calculator_keras.py]: It will probably be used with methods related to both once the weights part is implemented. I can remove it for now, but it's probably something that we will need in my next PR.

File: model_compression_toolkit/core/keras/keras_implementation.py [hunk 11]
```diff
@@ -458,39 +464,20 @@ def get_node_distance_fn(self, layer_class: type,
             return compute_cs
         return compute_mse
 
-    def model_grad(self,
```
[the remark this answers is not in the document]
Comment (reuvenperetz) [reply model_compression_toolkit/core/keras/keras_implementation.py:461]: Yes.

File: model_compression_toolkit/core/common/hessian/trace_hessian_config.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+from model_compression_toolkit.constants import HESSIAN_OUTPUT_ALPHA, HESSIAN_NUM_ITERATIONS
+
+
+class TraceHessianConfig:
+    """
+    Configuration for Hessian-based information computation, specifically for the trace of the Hessian.
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/hessian/trace_hessian_config.py:20]: Same here, not sure that the second part of the sentence is necessary

File: model_compression_toolkit/core/common/hessian/trace_hessian_config.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+    This class is used to define parameters for the approximation of the trace of the Hessian matrix.
+    It's important to note that this does not compute the actual Hessian but approximates some Hessian-based data.
+    The computation can be computationally heavy and is based on the trace of the Hessian.
+
+    """
+
+    def __init__(self,
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/hessian/trace_hessian_config.py:28]: This class is becoming unnecessary if we are going to get rid of the "alpha" parameter.
If it does not used in the whole "fetching"/"caching" mechanisms, then maybe consider to remove it completely and just add there to arguments to the calculator.

File: model_compression_toolkit/core/common/hessian/trace_hessian_service.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        self.trace_hessian_configuration = trace_hessian_configuration
+        self.fw_impl = fw_impl
+
+        self.trace_hessian_request_to_score_list = {}
+
+    def _clear_saved_hessian_info(self):
+        """Clears the saved info approximations."""
+        self.trace_hessian_request_to_score_list={}
+
+    def count_saved_info_of_request(self, hessian_request:TraceHessianRequest) -> int:
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/hessian/trace_hessian_service.py:64]: Add a comment regarding what it is counting (request is computed and saved per image yada yada yada..)

File: model_compression_toolkit/core/common/mixed_precision/sensitivity_evaluation.py [hunk 15]
```diff
@@ -187,28 +199,44 @@ def _build_models(self) -> Any:
 
     def _compute_gradient_based_weights(self) -> np.ndarray:
         """
-        Computes the gradient-based weights using the framework's model_grad method per batch of images.
-
-        Returns: A vector of weights, one for each interest point,
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/mixed_precision/sensitivity_evaluation.py:192]: Why did you remove the return from the description?

File: model_compression_toolkit/core/common/quantization/core_config.py [hunk 16]
```diff
@@ -26,19 +25,26 @@ class CoreConfig:
...
         self.mixed_precision_config = mixed_precision_config
         self.debug_config = debug_config
+        self.trace_hessian_cfg = trace_hessian_cfg
 
     @property
     def mixed_precision_enable(self):
         return self.mixed_precision_config is not None
+
+    @property
+    def init_hessian_service(self):
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/quantization/core_config.py:49]: change the name, it sounds like a method that initiate the hessian service

Comment (ofirgo) [reply model_compression_toolkit/core/common/quantization/core_config.py:49]: also add return type and description

File: model_compression_toolkit/core/pytorch/back2framework/model_gradients.py [hunk 17]
```diff
@@ -32,181 +32,181 @@
 from model_compression_toolkit.logger import Logger
```
Comment (ofirgo) [inline model_compression_toolkit/core/pytorch/back2framework/model_gradients.py:32]: Is this file still under editing? (all the code is commented-out)

File: model_compression_toolkit/core/common/hessian/trace_hessian_request.py [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+
+class TraceHessianMode(Enum):
+    """
+    Enum representing the mode for trace Hessian's trace information computation.
+
+    This determines whether the Hessian's trace approximation is computed w.r.t weights or w.r.t activations.
+    Note: This is not the actual Hessian's trace but an approximation.
+    """
+    WEIGHTS = 0         # Hessian approximation based on weights
+    ACTIVATIONS = 1     # Hessian approximation based on activations
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/hessian/trace_hessian_request.py:30]: Consider changing to "ACTIVATION" (no "S"), because this is like this is all our code (including enums).
For example: ActivationQauntizationMethod, all Activation quantizers, QuantizationTarget.Activation, etc.

**Review by ofirgo — CHANGES_REQUESTED** [review]
Review not completed, some comments in the meantime.,

File: model_compression_toolkit/core/common/quantization/core_config.py (hunk 16 above)
Comment (reuvenperetz) [reply model_compression_toolkit/core/common/quantization/core_config.py:49]: We don't use it eventually. I'll remove it.

File: model_compression_toolkit/core/common/hessian/trace_hessian_config.py (hunk 13 above)
Comment (reuvenperetz) [reply model_compression_toolkit/core/common/hessian/trace_hessian_config.py:28]: Is there no way we will need different params in the future? Something the user might want to configure?

File: model_compression_toolkit/ptq/keras/quantization_facade.py [hunk 19]
```diff
@@ -134,14 +134,15 @@ def keras_post_training_quantization_experimental(in_model: Model,
...
-        tg, bit_widths_config = core_runner(in_model=in_model,
-                                            representative_data_gen=representative_data_gen,
-                                            core_config=core_config,
-                                            fw_info=fw_info,
-                                            fw_impl=fw_impl,
-                                            tpc=target_platform_capabilities,
-                                            target_kpi=target_kpi,
-                                            tb_w=tb_w)
+        # Ignore returned hessian service as PTQ does not use it
+        tg, bit_widths_config, _ = core_runner(in_model=in_model,
```
Comment (ofirgo) [inline model_compression_toolkit/ptq/keras/quantization_facade.py:138]: Why not init a hessian service in this facade? (you do in the pytorch facade, and it might be relevant for parameters selection if I'm not mistaken)

File: tests/keras_tests/function_tests/test_model_gradients.py [hunk 20]
```diff
@@ -89,29 +92,40 @@ def model_with_output_replacements(input_shape):
     return keras.Model(inputs=inputs, outputs=outputs)
 
 
-def representative_dataset():
-    yield [np.random.randn(1, 8, 8, 3).astype(np.float32)]
+def representative_dataset(num_of_inputs=1):
+    yield [np.random.randn(1, 8, 8, 3).astype(np.float32)]*num_of_inputs
+
+def _get_normalized_hessian_trace_approx(graph, interest_points, keras_impl, alpha, num_of_inputs=1):
```
Comment (ofirgo) [inline tests/keras_tests/function_tests/test_model_gradients.py:98]: If this part is extracted from the model_gradients then it shouldn't be part of this test.
Changed the test's assertion to fit the expected results without normalization

File: tests/pytorch_tests/function_tests/model_gradients_test.py [hunk 21]
```diff
@@ -193,17 +195,51 @@ def run_test(self, seed=0):
 
         ipts = [n for n in graph.get_topo_sorted_nodes()]
         output_list = [ipts[-1]]
-        model_grads = pytorch_impl.model_grad(graph_float=graph,
-                                              model_input_tensors=input_tensors,
-                                              interest_points=ipts,
-                                              output_list=output_list,
-                                              all_outputs_indices=[len(ipts) - 1],
-                                              alpha=0)
+        # model_grads = pytorch_impl.model_grad(graph_float=graph,
```
Comment (ofirgo) [inline tests/pytorch_tests/function_tests/model_gradients_test.py:198]: remove commented-out code (in the entire file)

File: tests/pytorch_tests/function_tests/model_gradients_test.py [hunk 22]
```diff
@@ -12,6 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 # ==============================================================================
+import functools
```
Comment (ofirgo) [inline tests/pytorch_tests/function_tests/model_gradients_test.py:15]: couldn't fine where is it being used?

File: tests/pytorch_tests/function_tests/test_function_runner.py [hunk 23]
```diff
@@ -108,9 +108,8 @@ def test_model_gradients(self):
         ModelGradientsAdvancedModelTest(self).run_test()
         ModelGradientsMultipleOutputsTest(self).run_test()
         ModelGradientsOutputReplacementTest(self).run_test()
-        ModelGradientsMultipleOutputsModelTest(self).run_test()
+        # ModelGradientsMultipleOutputsModelTest(self).run_test()
```
Comment (ofirgo) [inline tests/pytorch_tests/function_tests/test_function_runner.py:111]: Add todo to remove comment or test

File: tests/pytorch_tests/function_tests/test_sensitivity_eval_output_replacement.py [hunk 24]
```diff
@@ -101,7 +107,11 @@ def __init__(self, unit_test):
 
     def run_test(self, seed=0, **kwargs):
         model = argmax_output_model()
-        self.verify_test_for_model(model)
+        with self.unit_test.assertRaises(Exception) as e:
```
Comment (ofirgo) [inline tests/pytorch_tests/function_tests/test_sensitivity_eval_output_replacement.py:110]: Can you explain this addition please?

File: tests/keras_tests/function_tests/test_hessian_service.py [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+                              name="bn1")(x)
+    outputs = ReLU()(x_bn)
+    return keras.Model(inputs=inputs, outputs=outputs)
+
+
+def representative_dataset(num_of_inputs=1):
+    yield [np.random.randn(2, 8, 8, 3).astype(np.float32)] * num_of_inputs
+
+
+class TestHessianService(unittest.TestCase):
```
Comment (ofirgo) [inline tests/keras_tests/function_tests/test_hessian_service.py:32]: Just verifying - did you include a test that asks for more images than what is already computed and triggers a computation after fetching the cached results? If not, I think this is an important one.

**Review by ofirgo — CHANGES_REQUESTED** [review]
Completed the review

**Review by ofirgo — APPROVED** [review]
